src/SignalR/common/Http.Connections.Common/src/NegotiateProtocol.cs

```diff
@@ -247,5 +338,157 @@ private static AvailableTransport ParseAvailableTransport(JsonTextReader reader)
+                        {
+                            throw new InvalidDataException($"Missing required property '{TransferFormatsPropertyName}'.");
+                        }
+
+                        return availableTransport;
+                    default:
+                        throw new InvalidDataException($"Unexpected token '{reader.TokenType}' when reading available transport JSON.");
+                }
+            }
+
+            throw new InvalidDataException("Unexpected end when reading JSON.");
+        }
+
+        private static bool CheckRead(ref Utf8JsonReader reader)
+        {
+            if (!reader.Read())
+            {
+                throw new InvalidDataException("Unexpected end when reading JSON.");
+            }
+
+            return true;
+        }
+
+        private static void EnsureObjectStart(ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new InvalidDataException($"Unexpected JSON Token Type '{GetTokenString(reader.TokenType)}'. Expected a JSON Object.");
+            }
+        }
+
+        private static string GetTokenString(JsonTokenType tokenType)
+        {
+            switch (tokenType)
+            {
+                case JsonTokenType.None:
+                    break;
+                case JsonTokenType.StartObject:
+                    return "StartObject";
+                case JsonTokenType.StartArray:
+                    return "StartArray";
+                case JsonTokenType.PropertyName:
+                    return "PropertyName";
+                default:
+                    break;
+            }
+            return tokenType.ToString();
+        }
+
+        private static void EnsureArrayStart(ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartArray)
+            {
+                throw new InvalidDataException($"Unexpected JSON Token Type '{GetTokenString(reader.TokenType)}'. Expected a JSON Array.");
+            }
+        }
+
+        private static bool ArrayEqual(ReadOnlySpan<byte> left, byte[] right)
+        {
+            if (left.Length != right.Length)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < left.Length; ++i)
+            {
+                if (left[i] != right[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static void Skip(ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType == JsonTokenType.PropertyName)
+            {
+                reader.Read();
+            }
+
+            if (reader.TokenType == JsonTokenType.StartObject || reader.TokenType == JsonTokenType.StartArray)
+            {
+                int depth = reader.CurrentDepth;
+                while (reader.Read() && depth < reader.CurrentDepth)
```

Comment:
This might need to be `depth <= reader.CurrentDepth`

At least that's what I observed in https://github.com/dotnet/core-setup/pull/5009/files#diff-02337f8b4f14d01b128b474cc5a5cf2aR98